File: examples/dekita/.claude/hooks/script_test_reminder.py
#!/usr/bin/env python3
"""PostToolUse hook: Remind to add tests when new functions are added to scripts.

Issue #1247: When editing .claude/scripts/*.py files and adding new function
definitions, this hook reminds the developer to add corresponding tests
if a test file already exists.
"""

# - Single responsibility: Remind to add tests when new functions are added
# - Uses systemMessage (non-blocking) as this is a reminder, not enforcement
# - Does not overlap with existing hooks

import json
import re
from pathlib import Path

from lib.execution import log_hook_execution
from lib.results import print_continue_and_log_skip
from lib.session import create_hook_context, parse_hook_input


def detect_new_functions(old_string: str, new_string: str) -> list[str]:
    """Detect newly added function definitions (including async functions).

    Args:
        old_string: Original code content
        new_string: New code content after edit

    Returns:
        List of newly added function names
    """
    # Match both 'def foo(' and 'async def foo('
    pattern = r"^(?:async\s+)?def\s+(\w+)\s*\("
    old_funcs = set(re.findall(pattern, old_string, re.MULTILINE))
    new_funcs = set(re.findall(pattern, new_string, re.MULTILINE))
    return sorted(new_funcs - old_funcs)


def find_test_file(script_path: str) -> Path | None:
    """Find corresponding test file for a script.

    Args:
        script_path: Path to the script file

    Returns:
        Path to test file if it exists, None otherwise
    """
    path = Path(script_path)
    # Convert hyphens to underscores for test file naming convention
    normalized_name = path.stem.replace("-", "_")
    test_file = path.parent / "tests" / f"test_{normalized_name}.py"
    return test_file if test_file.exists() else None


def main() -> None:
    """Main entry point for the hook."""
    result: dict = {"continue": True}

    try:
        input_data = parse_hook_input()
        # Issue #2607: Create context for session_id logging
        ctx = create_hook_context(input_data)
        tool_input = input_data.get("tool_input", {})

        file_path = tool_input.get("file_path", "")

        # Only target .claude/scripts/*.py files
        if ".claude/scripts/" not in file_path or not file_path.endswith(".py"):
            print_continue_and_log_skip("script-test-reminder", "not a script file", ctx=ctx)
            return

        # Exclude files in tests directory
        if "/tests/" in file_path:
            print_continue_and_log_skip("script-test-reminder", "test file excluded", ctx=ctx)
            return

        old_string = tool_input.get("old_string", "")
        new_string = tool_input.get("new_string", "")

        # Detect new function definitions
        new_functions = detect_new_functions(old_string, new_string)

        if not new_functions:
            print_continue_and_log_skip(
                "script-test-reminder", "no new functions detected", ctx=ctx
            )
            return

        # Check if corresponding test file exists
        test_file = find_test_file(file_path)

        if test_file:
            func_list = ", ".join(new_functions)
            result["systemMessage"] = (
                f"💡 テストファイルが存在します: {test_file}\n"
                f"   新しい関数を追加した場合、テストも追加してください。\n"
                f"   追加された関数: {func_list}"
            )
            log_hook_execution(
                "script-test-reminder",
                "remind",
                f"New functions: {func_list}",
                {"file": file_path, "test_file": str(test_file), "functions": new_functions},
            )
    except Exception:
        # Never fail the hook - just skip reminder
        pass

    print(json.dumps(result))


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/script_test_reminder.ts
#!/usr/bin/env bun
/**
 * PostToolUse hook: Remind to add tests when new functions are added to scripts.
 *
 * Why:
 *   When editing .claude/scripts/*.py files and adding new function
 *   definitions, this hook reminds the developer to add corresponding tests
 *   if a test file already exists.
 *
 * What:
 *   - Single responsibility: Remind to add tests when new functions are added
 *   - Uses systemMessage (non-blocking) as this is a reminder, not enforcement
 *   - Does not overlap with existing hooks
 *
 * Remarks:
 *   - PostToolUse:Edit hookとして発火
 *   - .claude/scripts/*.pyファイルへの編集時のみ対象
 *
 * Changelog:
 *   - silenvx/dekita#1247: フック追加
 *   - silenvx/dekita#2874: TypeScript移行
 */

import { existsSync } from "node:fs";
import { basename, dirname, join } from "node:path";
import { logHookExecution } from "../lib/logging";
import { createHookContext, parseHookInput } from "../lib/session";

const HOOK_NAME = "script-test-reminder";

/**
 * Detect newly added function definitions (including async functions and class methods).
 *
 * Uses occurrence counting to detect new methods even when a method with the same name
 * already exists (e.g., `__init__`, `run` in different classes).
 *
 * @param oldString - Original code content
 * @param newString - New code content after edit
 * @returns List of newly added function names (deduplicated and sorted)
 */
export function detectNewFunctions(oldString: string, newString: string): string[] {
  // Helper to extract all function names as an array (preserving duplicates)
  const getFuncs = (str: string): string[] => {
    const funcs: string[] = [];
    for (const match of str.matchAll(/^\s*(?:async\s+)?def\s+(\w+)\s*\(/gm)) {
      funcs.push(match[1]);
    }
    return funcs;
  };

  const oldFuncs = getFuncs(oldString);
  const newFuncs = getFuncs(newString);

  // Count occurrences in old string
  const oldCounts = new Map<string, number>();
  for (const func of oldFuncs) {
    oldCounts.set(func, (oldCounts.get(func) || 0) + 1);
  }

  // Track new occurrences and detect additions
  const added: string[] = [];
  const newCounts = new Map<string, number>();

  for (const func of newFuncs) {
    const count = (newCounts.get(func) || 0) + 1;
    newCounts.set(func, count);
    // If this occurrence count exceeds old count, it's a new addition
    if (count > (oldCounts.get(func) || 0)) {
      added.push(func);
    }
  }

  // Return unique names, sorted
  return [...new Set(added)].sort();
}

/**
 * Find corresponding test file for a script.
 *
 * @param scriptPath - Path to the script file
 * @returns Path to test file if it exists, null otherwise
 */
function findTestFile(scriptPath: string): string | null {
  // Convert hyphens to underscores for test file naming convention
  const stem = basename(scriptPath, ".py");
  const normalizedName = stem.replace(/-/g, "_");
  const testFile = join(dirname(scriptPath), "tests", `test_${normalizedName}.py`);
  return existsSync(testFile) ? testFile : null;
}

interface ContinueResult {
  continue: boolean;
  systemMessage?: string;
}

async function main(): Promise<void> {
  const result: ContinueResult = { continue: true };

  try {
    const inputData = await parseHookInput();
    const ctx = createHookContext(inputData);
    const toolInput = inputData.tool_input || {};
    const filePath = (toolInput as { file_path?: string }).file_path || "";

    // Only target .claude/scripts/*.py files
    if (!filePath.includes(".claude/scripts/") || !filePath.endsWith(".py")) {
      logHookExecution(HOOK_NAME, "approve", "not a script file", {
        session_id: ctx.sessionId,
      });
      console.log(JSON.stringify(result));
      return;
    }

    // Exclude files in tests directory
    if (filePath.includes("/tests/")) {
      logHookExecution(HOOK_NAME, "approve", "test file excluded", {
        session_id: ctx.sessionId,
      });
      console.log(JSON.stringify(result));
      return;
    }

    const oldString = (toolInput as { old_string?: string }).old_string || "";
    const newString = (toolInput as { new_string?: string }).new_string || "";

    // Detect new function definitions
    const newFunctions = detectNewFunctions(oldString, newString);

    if (newFunctions.length === 0) {
      logHookExecution(HOOK_NAME, "approve", "no new functions detected", {
        session_id: ctx.sessionId,
      });
      console.log(JSON.stringify(result));
      return;
    }

    // Check if corresponding test file exists
    const testFile = findTestFile(filePath);

    if (testFile) {
      const funcList = newFunctions.join(", ");
      result.systemMessage = `💡 テストファイルが存在します: ${testFile}\n   新しい関数を追加した場合、テストも追加してください。\n   追加された関数: ${funcList}`;
      logHookExecution(HOOK_NAME, "approve", `New functions: ${funcList}`, {
        file: filePath,
        test_file: testFile,
        functions: newFunctions,
      });
    }
  } catch {
    // Never fail the hook - just skip reminder
  }

  console.log(JSON.stringify(result));
}

if (import.meta.main) {
  main();
}
